arborist/service.go

```diff
@@ -0,0 +1,39 @@
+package arborist
+
+// Representation of a service that uses arborist for access control.
+type Service struct {
+	// The name for the service; should have uniqueness enforced by the
+	// `AuthEngine`.
+	ID string
+
+	// Services can update arborist with entries mapping from URIs to resources,
+	// so then arborist can be given just a URI in place of a resource ID and
+	// translate that into the appropriate resource.
+	uri_to_resource map[string]*Resource
```

Comment:
how is a service coupled with a map of resources?

Reply:
Yeah I'm not sure why this made sense at the time. Decoupling in progress

Reply:
The idea here was that arborist could keep track of a service mapping URLs to the actual resource instances, so if all the service has is `"/user/groups/ABC"` or something arborist knows the actual resource that represents.

Reply:
it seems to me that usually a service could be coupled with a list of actions that can be done. And for a type of resource, you can do a list of service.action... ?
eg: you can do ec2.RunInstance on resource: {type: vpc, name: dev}, and you can do metadata.download on resource: {type:project, name: bpa}

Reply:
so does that mean different service has different URIs to represent same resource? if not, we can just have one mapping from URI to Resource?

Reply:
so you're thinking the *actions* should be coupled to a service? and then also that the resources should have a type assigned? what do you mean be list of service.action for a type of resource?

Reply:
we could have just one mapping from URI to resource, just thinking about if two different services use the same relative path for something and both tried to register that path

Reply:
action is already composed of service+method+resource, right? but I am just thinking you can only have a list of predefined methods on particular service/application. Like it doesn't make sense to have method:uploaddata on service:jupyterhub, vice versa for method:startJupyterInstance on service:metadata/sheepdog

same logic for resource. you won't do {serivice:metadata, method:submit} on resource:user/A

Reply:
Sorry I typo’d in the first comment, s/action/method/g